shared/src/main/scala/mlscript/helpers.scala

```diff
@@ -6,6 +6,55 @@ import scala.collection.immutable.SortedSet
 
 import mlscript.utils._, shorthands._
 
+/**
+  * A class to hold to the context while converting mlscript types to
+  * typescript representation.
+  * 
+  * polarity tracks polarity of the current type.
+  * * a type has no polarity by default
+  * * a type has negative polarity at function input position
+  * * a type has positive polarity at function output position
+  * 
+  * argCounter tracks number of new argument variable names introduced
+  * and then is used to generate unique names
+  * 
+  * newTypeAlias and typeParams stores information of any new type aliases and params
+  * defined when converting. This information is collected to arrange the
+  * source code in proper order.
+  * 
+  * Existing type vars is used to maintain a mapping from originally
+  * created type variables to their names. This is then used to resolve
+  * `TypeVar` to their SourceCode representation.
+  *
+  * @param polarity
+  * @param argCounter
+  * @param existingTypeVars
+  * @param newTypeAlias
+  * @param typeParams
+  */
+class ToTsTypegenContext(
+  var polarity: Option[Boolean] = None,
```

Comment:
`polarity` should probably be immutable, and you should copy the context and pass it locally when going into a type subexpression. OTOH `argCounter` can remain a `var` but would be better as a local variable in `toTsTypeSourceCode`, as I tried to explain in https://github.com/hkust-taco/mlscript/issues/57#issuecomment-1027636268 (did you see the comment?)